### DISCUSSION TITLE
Using llama3.2 as local llm for deepeval

### DISCUSSION THREAD

**Discussion by lorenzofilitti** [confident-ai/deepeval#1203] [Q&A]
Hi I wanted to test deepeval. I set llama3.2:latest as my local model with the command: deepeval set-local-model --model-name=llama3.2:latest --base-url="http://localhost:11434/v1/" --api-key="ollama".

When I run the example code on deep eval website and I specify the model I want to use(llama3.2:latest) it doesn't find it. How can I solve this? The code:

```
from deepeval.metrics import GEval
from deepeval.test_case import LLMTestCaseParams

correctness_metric = GEval(
    name="Correctness",
    criteria="Determine whether the actual output is factually correct based on the expected output.",
    # NOTE: you can only provide either criteria or evaluation_steps, and not both
    evaluation_steps=[
        "Check whether the facts in 'actual output' contradicts any facts in 'expected output'",
        "You should also heavily penalize omission of detail",
        "Vague language, or contradicting OPINIONS, are OK"
    ],
    evaluation_params=[LLMTestCaseParams.INPUT, LLMTestCaseParams.ACTUAL_OUTPUT, 
    LLMTestCaseParams.EXPECTED_OUTPUT],
    model=llama3.2:latest
)
```

**Comment by antoninoLorenzo** [comment]
That's not clear how you are running the code, there are two approaches:
- using `deepeval` command
- manually running the code

I think that when you set a local LLM like this you should use the first approach.

Personally I suggest using the second approach for higher flexibility, however this involves writing a `DeepEvalBaseLLM` subclass that you pass as an instance to the `model` parameter; refer to [Using Custom LLMs for Evaluation](https://docs.confident-ai.com/docs/guides-using-custom-llms#).

**Comment by soyboy17** [comment]
Hey Lorenzo! I found I needed to set it up in a similar way to Google VertexAI on the [Metrics Introduction](https://docs.confident-ai.com/docs/metrics-introduction) page (with a few adjustments). Apologies for the formatting, but this is what worked for me

from deepeval.metrics import GEval
from deepeval.test_case import LLMTestCaseParams, LLMTestCase

# Modules for model implementation
from deepeval.models.base_model import DeepEvalBaseLLM
from langchain_ollama import ChatOllama

# Create a class to implement the local Ollama Llama model for DeepEval
class OllamaLlama(DeepEvalBaseLLM):
    def __init__(self, model):
        self.model = model
    
    def load_model(self):
        return self.model
    
    def generate(self, prompt: str) -> str:
        chat_model = self.load_model()
        return chat_model.invoke(prompt).content
    
    async def a_generate(self, prompt: str) -> str:
        chat_model = self.load_model()
        res = await chat_model.ainvoke(prompt)
        return res.content
    
    def get_model_name(self):
        return "llama3.1:latest"
    
# Config for the local Llama model
custom_model_llama = ChatOllama(
    model="llama3.1:latest",
    base_url="http://localhost:11434/",
    api_key="ollama"
)

# Instantiate the model
llama_model = OllamaLlama(custom_model_llama)

# Define your test case
test_case = LLMTestCase(
    input = "What if these shoes don't fit?",
    actual_output = "We offer a 30-day full refund at no extra cost.",
    expected_output = "You're eligible for a free full refund within 30 days of purchase.",
)

correctness_metric = GEval(
    name="Correctness",
    #criteria="Determine whether the actual output is factually correct based on the expected output.",
    # NOTE: you can only provide either criteria or evaluation_steps, and not both
    evaluation_steps=[
        "Check whether the facts in 'actual output' contradicts any facts in 'expected output'",
        "You should also heavily penalize omission of detail",
        "Vague language, or contradicting OPINIONS, are OK"
    ],
    evaluation_params=[LLMTestCaseParams.INPUT, LLMTestCaseParams.ACTUAL_OUTPUT, 
    LLMTestCaseParams.EXPECTED_OUTPUT],
    model=llama_model
)

# Print completeness score and reason
correctness_metric.measure(test_case)
print("Correctness: ", correctness_metric.score)
print("Reason: ", correctness_metric.reason)

**Reply by antoninoLorenzo** [reply]
Yes, that's what I was talking about (except the formatting 😭). 

As I said this approach is more flexible, for example I am working without a framework (like langchain) and with Ollama deployed on Google Colab; their support for custom LLMs made it possible for me to use deepeval as evaluation framework and in general it seems like the project owners actually care about usability, even for specific use cases.

**Comment by kritinv** [comment]
@lorenzofilitti @antoninoLorenzo @soyboy17 You can still define custom models for evaluations, but we've added a lot of native integrations with several model providers, so you can now use providers like Ollama out of the box without defining a custom class.

Integration docs: https://www.deepeval.com/integrations/models/openai

**Comment by vk6010290208-oss** [comment]
- _********_

**Comment by akashcodes23** [comment]
This is a common setup issue with Ollama + DeepEval. The problem is that deepeval set-local-model stores the config, but when you explicitly pass model= in your test case, it overrides the stored config and tries to find the model via the OpenAI client — which doesn't know about your local Ollama endpoint.
Fix: instead of passing model="llama3.2:latest" directly, use DeepEval's OllamaModel wrapper:
```
from deepeval.models import OllamaModel

ollama_model = OllamaModel(model="llama3.2:latest")

metric = AnswerRelevancyMetric(
    threshold=0.7,
    model=ollama_model
)
```
Make sure Ollama is running (ollama serve) before executing tests. Also double-check your model name with ollama list — it needs to match exactly.
This bypasses the CLI config entirely and directly hooks DeepEval into your local Ollama instance.

**Comment by kinthaiofficial** [comment]
Local LLM as evaluator is worth the setup cost — it takes the dependency on expensive frontier API calls out of your CI loop, which matters when you're running eval suites on every commit.

A few things to calibrate expectations on:

**Evaluator model size matters more than model family** — a 70B local model is generally a better evaluator than a 7B frontier model for nuanced quality judgments. The sweet spot for most evaluation tasks is 32B-70B: capable enough to understand what "good" looks like, fast enough to not bottleneck CI.

**Task-specific prompting over generic rubrics** — llama3.2 and similar models respond better to very specific evaluation criteria than to generic "rate from 1-10" prompts. For agent output evaluation, we found that decomposing into specific dimensions (did the agent complete the stated goal? did it stay within budget? did it cite sources when needed?) and evaluating each separately gives more reliable scores than holistic ratings.

**Calibration against frontier model** — before fully switching to local eval, run both local and frontier evaluators on 200 samples and measure agreement rate. For simple correctness checks, local models are 90%+ aligned. For nuanced quality, agreement can drop to 70-80%, which may or may not be acceptable depending on your use case.

**Consistency sampling** — for local models that are less deterministic than frontier models, sampling the same eval 3 times and averaging the scores (or taking majority vote for binary judgments) significantly improves reliability.

We use a similar local-eval setup for KinthAI's agent output quality checks: https://blog.kinthai.ai/221-agents-multi-agent-coordination-lessons covers the evaluation framework design.

What evaluation dimensions are you running deepeval for — factual correctness, format compliance, or something more qualitative?
